--- ./dwm.c	2022-12-09 20:47:43.418893156 +0100
+++ ./dwm.c	2022-12-09 20:47:46.992226389 +0100
@@ -238,6 +238,8 @@ static void setfocus(Client *c);
 static void setfullscreen(Client *c, int fullscreen);
 static void setlayout(const Arg *arg);
 static void setmfact(const Arg *arg);
+static void shiftview(const Arg *arg);
+static void shiftview_to_not_empty(const Arg *arg);
 static void setup(void);
 static void seturgent(Client *c, int urg);
 static void showhide(Client *c);
@@ -2005,6 +2007,37 @@ seturgent(Client *c, int urg)
 	XFree(wmh);
 }
 
+/** Function to shift the current view to the left/right
+ *
+ * @param: "arg->i" stores the number of tags to shift right (positive value)
+ *          or left (negative value)
+ */
+void
+shiftview(const Arg *arg) {
+       Arg shifted;
+
+       if(arg->i > 0) // left circular shift
+               shifted.ui = (selmon->tagset[selmon->seltags] << arg->i)
+                  | (selmon->tagset[selmon->seltags] >> (LENGTH(tags) - arg->i));
+
+       else // right circular shift
+               shifted.ui = selmon->tagset[selmon->seltags] >> (- arg->i)
+                  | selmon->tagset[selmon->seltags] << (LENGTH(tags) + arg->i);
+
+       view(&shifted);
+}
+
+void
+shiftview_to_not_empty(const Arg *arg) {
+  int i;
+  unsigned int tagset;
+  Client *c;
+  Arg shifted;
+
+  // function runs only when single tag is viewed
+  for (tagset = selmon->tagset[selmon->seltags]; tagset; tagset <<= 1){
+    if ((tagset & (TAGMASK - 1)) != 0 && (tagset & (TAGMASK - 1)) != tagset){
+      return;
+    }
+  }
+
+  i = arg->i;
+  tagset = selmon->tagset[selmon->seltags];
+
+  while (-(int)LENGTH(tags) < i && i < (int)LENGTH(tags)) {
+    if(arg->i > 0) { // left circular shift
+      shifted.ui = (tagset << i) | (tagset >> (LENGTH(tags) - i));
+      i++;
+    } else { // right circular shift
+      shifted.ui = tagset >> (- i) | tagset << (LENGTH(tags) + i);
+      i--;
+    }
+    
+    for (c = selmon->stack; c; c = c->snext) {
+      if (tagset & shifted.ui) {
+        view(&shifted);
+        return;
+      }
+    }
+  }
+  // if nothing is found
+  shiftview(arg);
+}
+
+
 void
 showhide(Client *c)
 {
